### DISCUSSION TITLE
Support for dynamic_reconfigure in Parameter panel

### DISCUSSION THREAD

**Discussion by rgov** [foxglove/community#178] [Q&A]
The Parameter panel does not appear to support [dynamic_reconfigure](http://wiki.ros.org/dynamic_reconfigure). This would be useful for tuning parameters for live systems.

The Parameter panel does show params that are "managed" by dynamic_reconfigure because it reflects them to the parameter server. But, when they are updated in Studio, it does not trigger the reconfigure callback to allow the node to react to the new value.

Also, dynamic_reconfigure gives information like documentation strings and min and max ranges which might be useful for presenting a UI as `rqt_reconfigure` does.

**Comment by rgov** [comment]
I think you would need to do these things:

1. Find topics of type [dynamic_reconfigure/ParamDescription](http://docs.ros.org/en/noetic/api/dynamic_reconfigure/html/msg/ParamDescription.html) and interpret the parameter descriptions to build a UI

2. Find services that match [this interface](http://docs.ros.org/en/noetic/api/dynamic_reconfigure/html/srv/Reconfigure.html), usually `/<node>/set_parameters`, and invoke the service whenever the UI updates, and use its return value to override what you display in the UI

3. Subscribe to `/<node>/parameter_updates` as well to update the UI.

[Relevant tutorial](http://wiki.ros.org/roslibjs/Tutorials/Writing%20a%20dynamic_reconfigure%20client%20with%20roslibjs) from roslib.js.

**Comment by DevnathNair** [comment]
This would be amazing. Would make it all the more functional.

**Comment by chenjunnn** [comment]
Please do add this feature to Studio, it is very useful for me too!

**Comment by jhurliman** [comment]
I just bumped into this setting up the `realsense2_camera` node which makes heavy use of dynamic_reconfigure for live tuning of the camera. Being able to tune in studio without launching rqt_reconfigure would be lovely.

**Comment by Tobias-Pe** [comment]
We are interested to moving our tool chain to a one-in-all solution and are looking at foxglove currently.

Do I understand from this discussion correctly that to continue using dynamic reconfigure, that I would still need rqt_gui ?

**Reply by jtbandes** [reply]
Yes, Foxglove Studio does not have built in support for dynamic_reconfigure.

Based on @rgov's description above I believe that someone might be able to implement most of this as a [panel extension](https://foxglove.dev/docs/studio/extensions/getting-started#custom-panels).
